python/ray/autoscaler/v2/utils.py

```diff
@@ -507,3 +509,23 @@ def _parse_pending(cls, state: AutoscalingState) -> List[NodeInfo]:
             )
 
         return pending_nodes
+
+
+def is_autoscaler_v2() -> bool:
+    """
+    Check if the autoscaler is v2 from reading GCS internal KV.
+
+    Returns:
+        is_v2: True if the autoscaler is v2, False otherwise.
+
+    Raises:
+        Exception: if GCS address could not be resolved (e.g. ray.init() not called)
+    """
+
+    # See src/ray/common/constants.h for the definition of this key.
+    return (
+        _internal_kv_get(
```

Comment:
Isn't this too expensive? why don't we just follow the previous pattern?

Reply:
More concretely, I am concerned about the future (that people overuse this API)

Reply:
maybe we can keep the state in memory after the first call?

Reply:
This method would be removed as soon as we migrate to autoscaler v2.

The callers right now are not in any critical path, mainly for observability and sdks.

Previous pattern is not ideal since client would also need to specify the env variable. (and it might be inconsistent with the server)

Reply:
The caching is a bit tricky, i tried caching in the module (having a global variable outside of the function), but this would interfere with multiple ray sessions (e.g. ray.init(v2) -> ray.shutdown() -> ray.init(v1), would have cache value. 

We could not cache it on the global ray driver also because one current usecase (in the reporter head) does not have driver initialized.

Reply:
can you comment me which APIs use this flag under the hood? Let's just make sure we don't have excessive calls here.

Reply:
Right now, 

1. debug_status (from `ray status`): https://github.com/ray-project/ray/blob/1a1fdeb68329839c225b7136c983689bae70525e/python/ray/autoscaler/_private/commands.py#L115-L127
2. debug_status (from reporter head's handling of `/api/cluster_events`): https://github.com/ray-project/ray/blob/3d4e4ab9636f7ad181b0f73c829699a5c3c1fe0b/dashboard/modules/reporter/reporter_head.py#L116
3. request_resources: https://github.com/ray-project/ray/blob/1a1fdeb68329839c225b7136c983689bae70525e/python/ray/autoscaler/_private/commands.py#L183-L185

Reply:
hmm the dashboard part is concerning because the internal kv get can block the thread for 30+ seconds in a large cluster. 

is there a way to make it async for that part?

Reply:
> but this would interfere with multiple ray sessions (e.g. ray.init(v2) -> ray.shutdown() -> ray.init(v1), would have cache value.

Also feel like this is not a problem? There's no use case you use autoscaler V1 in first ray init and V2 in the second ray.init within the same python script?

Reply:
> hmm the dashboard part is concerning because the internal kv get can block the thread for 30+ seconds in a large cluster.

The rest of the code (e.g. getting debug string) is already using internal kv with blocking. So it would already be bad. 

> Also feel like this is not a problem? There's no use case you use autoscaler V1 in first ray init and V2 in the second ray.init within the same python script?

Well, this happens in test. But I guess I will just rewrite the test to circumvent this.